splash/browser_tab.py

```diff
@@ -731,3 +729,95 @@ def log(self, message, min_level=None):
+        image = QImage(size, QImage.Format_ARGB32)
+        painter = QPainter(image)
+        web_page.mainFrame().render(painter)
+        return image
+
+    logger.log("png renderer: viewport too large (%s), using tiled rendering" %
+               size, min_level=2)
+    # One bug is worked around by rendering the page one tile at a time onto a
+    # small-ish temporary image.  The magic happens in viewport-window
+    # transformation:
+    #
+    # - Sizes of tile painter viewport and tile painter window match
+    #   webpage viewport size to avoid rescaling.
+    # - Tile painter window is moved appropriately so that tile region is
+    #   overlayed onto the temporary image.
+    tile_hsize = min(tile_maxsize, size.width())
+    tile_vsize = min(tile_maxsize, size.height())
+    htiles = 1 + (size.width() - 1) // tile_hsize
+    vtiles = 1 + (size.height() - 1) // tile_vsize
+    tile_image = QImage(QSize(tile_hsize, tile_vsize), QImage.Format_ARGB32)
+    tile_painter = QPainter(tile_image)
+    viewport = QRect(0, 0, size.width(), size.height())
+    tile_painter.setViewport(viewport)
+
+    # The other bug manifests itself when you do painter.drawImage trying to
+    # concatenate tiles onto a single image and once you reach 32'768 along
+    # either dimension all of a sudden drawImage simply stops drawing anything.
+    # The simplest workaround that comes to mind is to use pillow for pasting
+    # images.
+    pil_image = Image.new(mode='RGBA',
+                          size=(size.width(), size.height()))
+    for i in xrange(htiles):
+        for j in xrange(vtiles):
+            left, top = i * tile_hsize, j * tile_vsize
+            tile_painter.setWindow(viewport.translated(left, top))
+            logger.log("Rendering with window=%s" % tile_painter.window(),
+                       min_level=2)
+            web_page.mainFrame().render(tile_painter)
+            pil_tile_image = _qimage_to_pil_image(tile_image)
+
+            logger.log("Pasting rendered tile to coords: %s" %
+                       ((left, top),),
+                       min_level=2)
+            pil_image.paste(pil_tile_image, (left, top))
+    tile_painter.end()
+    return _pil_image_to_qimage(pil_image)
+
+
+# XXX: brain-dead simple endianness detection, is there a better way?
+_is_little_endian = \
+    struct.pack('<i', 0x11223344) != struct.pack('=i', 0x11223344)
+
+
+def _qimage_to_pil_image(qimage):
+    """Convert QImage (in ARGB32 format) to PIL.Image (in RGBA mode)."""
+    # In our case QImage uses 0xAARRGGBB format stored in host endian order,
+    # we must convert it to [0xRR, 0xGG, 0xBB, 0xAA] sequences used by pillow.
+    buf = qimage.bits().asstring(qimage.numBytes())
+    if not _is_little_endian:
+        # If host is big-endian, convert to little-endian first.
+        count = len(buf) // 4
+        buf = struct.pack('<%di' % count, *struct.unpack('=%di' % count, buf))
```

Comment:
It is documented as follows [here](http://qt-project.org/doc/qt-4.8/qimage.html#image-formats):

*... the QRgb typedef is equivalent to an unsigned int containing an ARGB quadruplet on the format 0xAARRGGBB.*
*32-bit images have no color table; instead, each pixel contains an QRgb value. There are three different types of 32-bit images storing RGB (i.e. 0xffRRGGBB), ARGB and premultiplied ARGB values respectively....*

Yes, we can and it is indeed faster. We'll lose exotic middle-endian support, but it should be ok. Changed.